### PR TITLE
chore(history): add e2e tests

### PR DESCRIPTION
**Pull request by markkaylor**
### What does it do?

- Add a `features.js` file to the app-template to enable future flags. Note: templates do not accept the `config` folder so it cannot be added directly when generating the app from template.
- Update run-e2e-tests script to write the `features.js` file to the test app so they run with future flags. 
- Write e2e tests for collection types and single types for history versions after creating and updating an entry. Publish is omitted for now. See below.

### Why is it needed?

- To setup e2e tests so we can test content-history as we develop
- To test what has been developed so far for content-history

### How to test it?

```
STRAPI_FEATURES_FUTURE_CONTENT_HISTORY=true yarn test:e2e --setup --domains=content-manager --concurrency=1
```

Run in UI or Debug mode by appending:

`-- -- ui` or `-- --debug`

### Known issues
There are a couple issues with the publish action that I believe may be resolved after draft and publish v2 is merged. The issues could also be related to this [TODO](https://github.com/strapi/strapi/blob/823857733be05ab7457d2fee8deccb4c1d7d78d0/packages/core/content-manager/server/src/history/services/history.ts#L24-L25). In any case I think these issues should be handled in another PR and then update the publish tests.
- The result in the [documents middleware](https://github.com/strapi/strapi/blob/823857733be05ab7457d2fee8deccb4c1d7d78d0/packages/core/content-manager/server/src/history/services/history.ts#L52) returns a different shape than the other actions (create and update) so accessing the `documentId` is not consistent across actions which then leaves us with `relatedDocumentId: null` on the history version.
- The middleware `context` for publish action is also different and does not contain  a `status` property which also leaves us with `status: null` on the history version.

### CODE REVIEW

File: e2e/constants.js
Comment (markkaylor) [inline e2e/constants.js:14]: ℹ️ Otherwise the history-versions are not cleared when db is reset

File: e2e/tests/content-history/history.spec.ts
Comment (remidej) [inline e2e/tests/content-history/history.spec.ts:99]: needs empty line above

File: e2e/tests/content-history/history.spec.ts
Comment (remidej) [inline e2e/tests/content-history/history.spec.ts:9]: I would add one more important aspect: checking that the history is isolated between the different locales.

At the end of the test, switch to a different locale, create something, save, go to history, and check that we only see this change on the new locale and not the stuff we did before on english content

**Review by remidej — COMMENTED** [review]
Tested locally ✅

File: e2e/tests/content-history/history.spec.ts
Comment (remidej) [inline e2e/tests/content-history/history.spec.ts:35]: Could you also test that you're auto redirected to `&id=versionId`? then you can delete [this file](https://github.com/strapi/strapi/blob/7463352479ed8ceab21a1a3ad6e16d1a528ad012/packages/core/admin/admin/src/content-manager/history/pages/tests/History.test.tsx#L2)

File: e2e/tests/content-history/history.spec.ts
Comment (markkaylor) [reply e2e/tests/content-history/history.spec.ts:9]: Good idea! 👍

File: e2e/tests/content-history/history.spec.ts
Comment (markkaylor) [reply e2e/tests/content-history/history.spec.ts:35]: Yes I'll take a look. That also reminds me I will update the waitForUrls to use a regex Josh sent me

File: e2e/app-template/features.js
Comment (joshuaellis) [reply e2e/app-template/features.js:3]: Do templates copy their configs? We could just add the config if so.

Comment (remidej) [reply e2e/app-template/features.js:3]: Templates were designed to just reuse the data structure, so we cherry picked the files and folders that made sense for it. I don't remember what the risks were for a more permissive approach though. We can change it if needed

File: e2e/tests/content-history/history.spec.ts
Comment (markkaylor) [reply e2e/tests/content-history/history.spec.ts:9]: Currently Single types throw an error when creating and entry in another locale so I'll only add the test for collection types right now and add a TODO for single types

File: e2e/app-template/features.js
Comment (joshuaellis) [reply e2e/app-template/features.js]: Might be worth throwing a message in slack to the wider group to see if anyone sees any limitations // complications to doing so. To me it feels like a good round addition to a template, for example if your template included plugins you configured for users.

Comment (markkaylor) [reply e2e/app-template/features.js]: Sounds good, will do. I think what's a bit tricky about the config folder is some things do make sense for a template like plugins as you suggested, but on the database config for example I'm not so sure about. 

But I agree we could probably include the config some or all of the config folder in the template export somehow. 

I'd prefer to do that in another PR targeting v4 as an enhancement and then pull it into v5. Already I think I'm going to have to backport some of the work I've done here that makes it easier to run tests with env variables.

Comment (joshuaellis) [reply e2e/app-template/features.js]: I think separate PR is understandable, dunno if we need it in v4. Trying to wind down feature dev. I think when releases is done we'll only be doing fixes and security.

But totally up to you!

File: examples/getstarted/config/features.js
Comment (markkaylor) [inline examples/getstarted/config/features.js:3]: oops

File: docs/docs/guides/e2e/00-setup.md
Comment (joshuaellis) [inline docs/docs/guides/e2e/00-setup.md:1]: Great updating ⭐

File: e2e/tests/content-history/history.spec.ts
Comment (joshuaellis) [inline e2e/tests/content-history/history.spec.ts:1]: is there a reason this isn't just `history` in the cm folder considering it's a feature of the CM and not it's own domain? :)

Comment (markkaylor) [reply e2e/tests/content-history/history.spec.ts:1]: Good catch, I'll move it

File: packages/core/admin/admin/src/content-manager/history/components/VersionHeader.tsx
Comment (markkaylor) [reply packages/core/admin/admin/src/content-manager/history/components/VersionHeader.tsx:26]: params.singleStypes returns => 'single-type, params.collectionType returns => 'collection-types'

Screenshot 2024-02-22 at 09 19 16

Comment (joshuaellis) [reply packages/core/admin/admin/src/content-manager/history/components/VersionHeader.tsx:26]: ah okay, do you see an advantage in having two variables? We just went for `collectionType` in the CM and discern which string it is 🤔 not that it really matters, just want to understand your thinking 😄

Comment (markkaylor) [reply packages/core/admin/admin/src/content-manager/history/components/VersionHeader.tsx:26]: 🤔 collectionType is undefined on a single-type page.

I believe the values come from here:

https://github.com/strapi/strapi/blob/85ca1e8e529ea27cc2abaac9cb25d969dad99767/packages/core/admin/admin/src/content-manager/routes.tsx#L34-L51

...but let me know if I missed something.

Side note:
I think it could be interesting to have a common variable for both, but I would prefer `:contentType` over `:collectionType`, and personally I would also prefer `:contentTypeUid` over `:slug`.

**Review by remidej — APPROVED** [review]
good to go if the CI passes :)

File: packages/core/admin/admin/src/content-manager/history/components/VersionHeader.tsx
Comment (joshuaellis) [reply packages/core/admin/admin/src/content-manager/history/components/VersionHeader.tsx:26]: ah I've seen what's happened, your values are coming from

https://github.com/strapi/strapi/blob/85ca1e8e529ea27cc2abaac9cb25d969dad99767/packages/core/admin/admin/src/content-manager/history/routes.tsx#L7-L28

You could roll it into one and reconfigure but it's really up to you if you care enough, it depends how annoying it is having to check different params etc. but I don't think it would be hard... It's a bit strange to be honest that you've set a parameter as the first part of the URL, I would have probably just done `useMatch` to figure it out....

RE. renaming I would love to and i thought about renaming `slug` to be `model` because that's what its referred to on the BE but it would be huge breaking changes for everyone in V5 who looks at the params and it felt a bit unnecessary. I think we could line it up for V6 where hopefully we won't be doing as large breaking changes. I agree its shit naming though 😄

Comment (markkaylor) [reply packages/core/admin/admin/src/content-manager/history/components/VersionHeader.tsx]: pushed an update to use useMatch, lmk if it's what you had in mind.

**Comment by markkaylor** [comment]
This PR is blocked until DTS can handle the models API similar to contentTypes

File: e2e/tests/content-releases/releases-page.spec.ts
Comment (joshuaellis) [inline e2e/tests/content-releases/releases-page.spec.ts:11]: These tests do pass on `main` though, why don't they pass now?

Comment (markkaylor) [reply e2e/tests/content-releases/releases-page.spec.ts]: This was a mistake, they should pass here too. Thanks and good catch.

**Comment by markkaylor** [comment]
Don't I'll squash 🙃

**Review by remidej — APPROVED** [review]

**Review by joshuaellis — APPROVED** [review]

**Pull request merged**
